File: src/pivot_array_2161.rs
// https://leetcode.com/problems/partition-array-according-to-given-pivot

pub fn pivot_array(nums: Vec<i32>, pivot: i32) -> Vec<i32> {
    let mut lesser = 0;
    let mut equal = 0;
    for &n in &nums {
        if n == pivot {
            equal += 1;
        } else if n < pivot {
            lesser += 1;
        }
    }

    let mut result = vec![pivot; nums.len()];
    let (less_slots, rest) = result.split_at_mut(lesser);
    let (_pivots, greater_slots) = rest.split_at_mut(equal);
    let mut less_slots = less_slots.iter_mut();
    let mut greater_slots = greater_slots.iter_mut();

    for n in nums {
        if n < pivot {
            *less_slots.next().unwrap() = n;
        } else if n > pivot {
            *greater_slots.next().unwrap() = n;
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::pivot_array;

    #[test]
    fn official1() {
        let nums = vec![9, 12, 5, 10, 14, 3, 10];
        let pivot = 10;
        let expected = vec![9, 5, 3, 10, 10, 12, 14];

        assert_eq!(expected, pivot_array(nums, pivot));
    }

    #[test]
    fn official2() {
        let nums = vec![-3, 4, 3, 2];
        let pivot = 2;
        let expected = vec![-3, 2, 4, 3];

        assert_eq!(expected, pivot_array(nums, pivot));
    }
}

File: cpp/src/pivot_array_2161.cpp
// https://leetcode.com/problems/partition-array-according-to-given-pivot

#include "test.hpp"

#include <algorithm>
#include <cstddef>
#include <iterator>
#include <vector>

    std::vector<int> pivotArray(std::vector<int> nums, int pivot) {
        const auto equal = static_cast<std::size_t>(std::ranges::count(nums, pivot));

        std::vector<int> result;
        result.reserve(nums.size());
        std::ranges::copy_if(nums, std::back_inserter(result), [&](int n) { return n < pivot; });
        result.insert(result.end(), equal, pivot);
        std::ranges::copy_if(nums, std::back_inserter(result), [&](int n) { return n > pivot; });
        return result;
    }

TEST(official1) {
    check_eq({9, 5, 3, 10, 10, 12, 14}, pivotArray({9, 12, 5, 10, 14, 3, 10}, 10));
}

TEST(official2) {
    check_eq({-3, 2, 4, 3}, pivotArray({-3, 4, 3, 2}, 2));
}
